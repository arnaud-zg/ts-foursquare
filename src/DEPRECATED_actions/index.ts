import { TLifeAction } from './life'

export type TRootAction = TLifeAction

export * from './life'
