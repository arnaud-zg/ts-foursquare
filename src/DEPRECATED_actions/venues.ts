import { ActionType, createAsyncAction } from 'typesafe-actions'
import { NRequest } from '../../types'
import {
  adaptGetVenuesCategories,
  adaptGetVenuesExplore,
  adaptGetVenuesLikes,
  adaptGetVenuesListed,
  adaptGetVenuesNextVenues,
  adaptGetVenuesSearch,
  adaptGetVenuesSimilar,
  adaptGetVenuesSuggestCompletion,
  adaptGetVenuesTrending,
} from '../adapter/venues'

export const getVenuesCategoriesAsync = createAsyncAction(
  'GET_VENUES_CATEGORIES_REQUEST',
  'GET_VENUES_CATEGORIES_SUCCESS',
  'GET_VENUES_CATEGORIES_FAILURE',
  'GET_VENUES_CATEGORIES_CANCEL'
)<void, ReturnType<typeof adaptGetVenuesCategories>, Error, string>()

export const getVenuesExploreAsync = createAsyncAction(
  'GET_VENUES_EXPLORE_REQUEST',
  'GET_VENUES_EXPLORE_SUCCESS',
  'GET_VENUES_EXPLORE_FAILURE',
  'GET_VENUES_EXPLORE_CANCEL'
)<
  NRequest.TVenuesExplorePayload,
  ReturnType<typeof adaptGetVenuesExplore>,
  Error,
  string
>()

export const getVenuesLikesAsync = createAsyncAction(
  'GET_VENUES_LIKES_REQUEST',
  'GET_VENUES_LIKES_SUCCESS',
  'GET_VENUES_LIKES_FAILURE',
  'GET_VENUES_LIKES_CANCEL'
)<
  NRequest.IVenuesLikesPayload,
  ReturnType<typeof adaptGetVenuesLikes>,
  Error,
  string
>()

export const getVenuesListedAsync = createAsyncAction(
  'GET_VENUES_LISTED_REQUEST',
  'GET_VENUES_LISTED_SUCCESS',
  'GET_VENUES_LISTED_FAILURE',
  'GET_VENUES_LISTED_CANCEL'
)<
  NRequest.IVenuesListedPayload,
  ReturnType<typeof adaptGetVenuesListed>,
  Error,
  string
>()

export const getVenuesNextVenuesAsync = createAsyncAction(
  'GET_VENUES_NEXT_VENUES_REQUEST',
  'GET_VENUES_NEXT_VENUES_SUCCESS',
  'GET_VENUES_NEXT_VENUES_FAILURE',
  'GET_VENUES_NEXT_VENUES_CANCEL'
)<
  NRequest.IVenuesNextVenuesPayload,
  ReturnType<typeof adaptGetVenuesNextVenues>,
  Error,
  string
>()

export const getVenuesSearchAsync = createAsyncAction(
  'GET_VENUES_SEARCH_REQUEST',
  'GET_VENUES_SEARCH_SUCCESS',
  'GET_VENUES_SEARCH_FAILURE',
  'GET_VENUES_SEARCH_CANCEL'
)<
  NRequest.TVenuesSearchPayload,
  ReturnType<typeof adaptGetVenuesSearch>,
  Error,
  string
>()

export const getVenuesSimilarAsync = createAsyncAction(
  'GET_VENUES_SIMILAR_REQUEST',
  'GET_VENUES_SIMILAR_SUCCESS',
  'GET_VENUES_SIMILAR_FAILURE',
  'GET_VENUES_SIMILAR_CANCEL'
)<
  NRequest.IVenuesSimilarPayload,
  ReturnType<typeof adaptGetVenuesSimilar>,
  Error,
  string
>()

export const getVenuesSuggestCompletionAsync = createAsyncAction(
  'GET_VENUES_SUGGEST_COMPLETION_REQUEST',
  'GET_VENUES_SUGGEST_COMPLETION_SUCCESS',
  'GET_VENUES_SUGGEST_COMPLETION_FAILURE',
  'GET_VENUES_SUGGEST_COMPLETION_CANCEL'
)<
  NRequest.TVenuesSuggestCompletionPayload,
  ReturnType<typeof adaptGetVenuesSuggestCompletion>,
  Error,
  string
>()

export const getVenuesTrendingAsync = createAsyncAction(
  'GET_VENUES_TRENDING_REQUEST',
  'GET_VENUES_TRENDING_SUCCESS',
  'GET_VENUES_TRENDING_FAILURE',
  'GET_VENUES_TRENDING_CANCEL'
)<
  NRequest.TVenuesTrendingPayload,
  ReturnType<typeof adaptGetVenuesTrending>,
  Error,
  string
>()

export const venuesActions = {
  getVenuesCategoriesAsync,
  getVenuesExploreAsync,
  getVenuesLikesAsync,
  getVenuesListedAsync,
  getVenuesNextVenuesAsync,
  getVenuesSearchAsync,
  getVenuesSimilarAsync,
  getVenuesSuggestCompletionAsync,
  getVenuesTrendingAsync,
}

export type TVenuesAction = ActionType<typeof venuesActions>
