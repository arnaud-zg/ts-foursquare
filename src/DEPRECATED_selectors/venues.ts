import { NStore, NVenue } from '../../types'

export const venueCategoriesStateSelector = (
  state: NStore.IState,
  venueId: string
): NVenue.ICategory[] => {
  const venueEntity = venuesEntitiesSelector(state)[venueId]
  return venueEntity ? venueEntity.categories : []
}

export const venueLocationSelector = (
  state: NStore.IState,
  venueId: string
): NVenue.ILocation | null => {
  const venueEntity = venuesEntitiesSelector(state)[venueId]
  return venueEntity ? venueEntity.location : null
}

export const venuesCategoriesSelector = (state: NStore.IState) =>
  venuesSelector(state).categories

export const venueSelector = (state: NStore.IState, venueId: string) =>
  venuesEntitiesSelector(state)[venueId]

export const venuesEntitiesSelector = (state: NStore.IState) =>
  venuesSelector(state).entities

export const venuesNextVenuesSelector = (state: NStore.IState) =>
  venuesSelector(state).nextVenues

export const venuesSimilarSelector = (state: NStore.IState) =>
  venuesSelector(state).similarVenues

export const venuesRecommendedPlacesSelector = (state: NStore.IState) =>
  venuesSelector(state).recommendedPlaces

export const venuesListedSelector = (state: NStore.IState) =>
  venuesSelector(state).venuesListed

export const venuesListedGroupSelector = (
  state: NStore.IState,
  groupKey: string
) => venuesSelector(state).venuesListed[groupKey]

export const venuesSelector = (state: NStore.IState) => state.venues
