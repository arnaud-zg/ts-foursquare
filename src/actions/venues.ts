import { ActionType, createAsyncAction } from 'typesafe-actions'
import { NRequest } from '../../types/request.d'
import { NVenue } from '../../types/venue.d'

interface IAsyncActionNameMapping {
  [actionType: string]: string
}

export enum EVenuesAction {
  GET_VENUES_SEARCH_REQUEST = 'GET_VENUES_SEARCH_REQUEST',
  GET_VENUES_SEARCH_SUCCESS = 'GET_VENUES_SEARCH_SUCCESS',
  GET_VENUES_SEARCH_FAILURE = 'GET_VENUES_SEARCH_FAILURE',
  GET_VENUES_SEARCH_CANCEL = 'GET_VENUES_SEARCH_CANCEL',
}

export const ASYNC_ACTION_NAME_MAPPING: IAsyncActionNameMapping = {
  [EVenuesAction.GET_VENUES_SEARCH_REQUEST]: 'getVenuesSearchAsync',
  [EVenuesAction.GET_VENUES_SEARCH_SUCCESS]: 'getVenuesSearchAsync',
  [EVenuesAction.GET_VENUES_SEARCH_FAILURE]: 'getVenuesSearchAsync',
  [EVenuesAction.GET_VENUES_SEARCH_CANCEL]: 'getVenuesSearchAsync',
}

export const getVenuesSearchAsync = createAsyncAction(
  'GET_VENUES_SEARCH_REQUEST',
  'GET_VENUES_SEARCH_SUCCESS',
  'GET_VENUES_SEARCH_FAILURE',
  'GET_VENUES_SEARCH_CANCEL'
)<NRequest.TVenuesSearchPayload, NVenue.IVenue[], Error, string>()

export const actions = {
  getVenuesSearchAsync,
}

export type TVenuesAction = ActionType<typeof actions>
