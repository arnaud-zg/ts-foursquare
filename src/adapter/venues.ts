import { NHttpStatuses } from 'http-response-status'
import { NPayload } from '../../types/payload.d'
import { NRecommendedPlaces } from '../../types/recommendedPlaces'
import { NVenue } from '../../types/venue.d'

export const adaptGetVenuesSearch = (
  payload: NPayload.IPayload<NVenue.IResponse>
): NVenue.IVenue[] =>
  !!payload.meta &&
  payload.meta.code === NHttpStatuses.ESuccess.OK &&
  !!payload.response &&
  !!payload.response.venues &&
  !!payload.response.venues.length
    ? payload.response.venues
    : []

export const adaptGetVenuesExplore = (
  payload: NPayload.IPayload<NRecommendedPlaces.IResponse>
): NRecommendedPlaces.IGroupItem[] =>
  !!payload.meta &&
  payload.meta.code === NHttpStatuses.ESuccess.OK &&
  !!payload.response &&
  !!payload.response.groups &&
  !!payload.response.groups.length &&
  !!payload.response.groups[0].items &&
  !!payload.response.groups[0].items.length
    ? payload.response.groups[0].items
    : []

export const adaptGetVenuesTrending = (
  payload: NPayload.IPayload<NVenue.IResponse>
): NVenue.IVenue[] => adaptGetVenuesSearch(payload)
