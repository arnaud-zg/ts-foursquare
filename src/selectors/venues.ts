import { NStore } from '../../types/store'

export const venuesSelector = (state: NStore.IState) => state.venues.entities

export const venuesCategoriesSelector = (state: NStore.IState) =>
  venuesSelector(state).categories || []

export const venueSelector = (state: NStore.IState, venueId: string) =>
  venuesSelector(state)[venueId]

export const venueCategoriesStateSelector = (
  state: NStore.IState,
  venueId: string
) => venuesSelector(state)[venueId].categories

export const venueLocationSelector = (state: NStore.IState, venueId: string) =>
  venuesSelector(state)[venueId].location
