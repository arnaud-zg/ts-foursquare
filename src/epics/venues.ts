import { Epic } from 'redux-observable'
import { of } from 'rxjs/internal/observable/of'
import { catchError, filter, map, switchMap, takeUntil } from 'rxjs/operators'
import { isActionOf } from 'typesafe-actions'
import { NStore } from '../../types/store'
import { TRootAction } from '../actions'
import {
  getVenuesCategoriesAsync,
  getVenuesExploreAsync,
  getVenuesLikesAsync,
  getVenuesNextVenuesAsync,
  getVenuesSearchAsync,
  getVenuesSimilarAsync,
  getVenuesSuggestCompletionAsync,
  getVenuesTrendingAsync,
} from '../actions/venues'
import {
  adaptGetVenuesCategories,
  adaptGetVenuesExplore,
  adaptGetVenuesLikes,
  adaptGetVenuesNextVenues,
  adaptGetVenuesSearch,
  adaptGetVenuesSimilar,
  adaptGetVenuesSuggestCompletion,
  adaptGetVenuesTrending,
} from '../adapter/venues'
import {
  getObservableVenuesCategories,
  getObservableVenuesExplore,
  getObservableVenuesLikes,
  getObservableVenuesNextVenues,
  getObservableVenuesSearch,
  getObservableVenuesSimilar,
  getObservableVenuesSuggestCompletion,
  getObservableVenuesTrending,
} from '../services/venues'

export const getVenuesCategoriesEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesCategoriesAsync.request)),
    switchMap(() =>
      getObservableVenuesCategories({ state$ }).pipe(
        map(adaptGetVenuesCategories),
        map(getVenuesCategoriesAsync.success),
        catchError(err => of(getVenuesCategoriesAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesCategoriesAsync.cancel)))
        )
      )
    )
  )

export const getVenuesExploreEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesExploreAsync.request)),
    switchMap(action =>
      getObservableVenuesExplore({ action, state$ }).pipe(
        map(adaptGetVenuesExplore),
        map(getVenuesExploreAsync.success),
        catchError(err => of(getVenuesExploreAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesExploreAsync.cancel)))
        )
      )
    )
  )

export const getVenuesLikesEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesLikesAsync.request)),
    switchMap(action =>
      getObservableVenuesLikes({ action, state$ }).pipe(
        map(adaptGetVenuesLikes),
        map(getVenuesLikesAsync.success),
        catchError(err => of(getVenuesLikesAsync.failure(err))),
        takeUntil(action$.pipe(filter(isActionOf(getVenuesLikesAsync.cancel))))
      )
    )
  )

export const getVenuesNextVenuesEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesNextVenuesAsync.request)),
    switchMap(action =>
      getObservableVenuesNextVenues({ action, state$ }).pipe(
        map(adaptGetVenuesNextVenues),
        map(getVenuesNextVenuesAsync.success),
        catchError(err => of(getVenuesNextVenuesAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesNextVenuesAsync.cancel)))
        )
      )
    )
  )

export const getVenuesSearchEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesSearchAsync.request)),
    switchMap(action =>
      getObservableVenuesSearch({ action, state$ }).pipe(
        map(adaptGetVenuesSearch),
        map(getVenuesSearchAsync.success),
        catchError(err => of(getVenuesSearchAsync.failure(err))),
        takeUntil(action$.pipe(filter(isActionOf(getVenuesSearchAsync.cancel))))
      )
    )
  )

export const getVenuesSimilarEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesSimilarAsync.request)),
    switchMap(action =>
      getObservableVenuesSimilar({ action, state$ }).pipe(
        map(adaptGetVenuesSimilar),
        map(getVenuesSimilarAsync.success),
        catchError(err => of(getVenuesSimilarAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesSimilarAsync.cancel)))
        )
      )
    )
  )

export const getVenuesTrendingEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesTrendingAsync.request)),
    switchMap(action =>
      getObservableVenuesTrending({ action, state$ }).pipe(
        map(adaptGetVenuesTrending),
        map(getVenuesTrendingAsync.success),
        catchError(err => of(getVenuesTrendingAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesTrendingAsync.cancel)))
        )
      )
    )
  )

export const getVenuesSuggestCompletionEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesSuggestCompletionAsync.request)),
    switchMap(action =>
      getObservableVenuesSuggestCompletion({ action, state$ }).pipe(
        map(adaptGetVenuesSuggestCompletion),
        map(getVenuesSuggestCompletionAsync.success),
        catchError(err => of(getVenuesSuggestCompletionAsync.failure(err))),
        takeUntil(
          action$.pipe(
            filter(isActionOf(getVenuesSuggestCompletionAsync.cancel))
          )
        )
      )
    )
  )
