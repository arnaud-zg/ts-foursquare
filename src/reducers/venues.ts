import { createReducer } from 'typesafe-actions'
import { NVenues } from '../../types/venues'
import { EVenuesAction, TVenuesAction } from '../actions/venues'

export const initialState: NVenues.IVenuesState = {
  venues: {},
}

export const venuesReducer = createReducer<NVenues.IVenuesState, TVenuesAction>(
  initialState,
  {
    [EVenuesAction.RESOLVE_GET_VENUES_SEARCH]: (state, action) => ({
      ...state,
      ...action.payload,
    }),
  }
)
