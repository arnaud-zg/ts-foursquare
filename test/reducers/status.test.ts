import { NHttpStatuses } from 'http-response-status'
import { getVenuesSearchAsync } from '../../src/DEPRECATED_actions/venues'
import { adaptGetVenuesSearch } from '../../src/adapter/venues'
import {
  initialState,
  statusReducer,
} from '../../src/DEPRECATED_reducers/status'
import { payload } from '../epics/__mocks__/getVenuesSearchAsync.resolve'
require('isomorphic-fetch')

describe('epics/status', () => {
  test.each`
    initialState    | action
    ${initialState} | ${getVenuesSearchAsync.request({ ll: '40.7099,-73.9622' })}
    ${initialState} | ${getVenuesSearchAsync.failure(new Error(`Error ${NHttpStatuses.EClientError.NOT_FOUND}`))}
    ${initialState} | ${getVenuesSearchAsync.cancel('Action cancelled')}
    ${initialState} | ${getVenuesSearchAsync.success(adaptGetVenuesSearch(payload))}
  `('should set isLoading property', ({ initialState, action }) => {
    expect(statusReducer(initialState, action)).toMatchSnapshot()
  })
})
