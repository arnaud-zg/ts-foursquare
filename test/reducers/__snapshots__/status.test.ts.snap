// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`epics/status should set isLoading property 1`] = `
Object {
  "getVenuesSearchAsync": Object {
    "hasError": false,
    "isLoading": true,
  },
}
`;

exports[`epics/status should set isLoading property 2`] = `
Object {
  "getVenuesSearchAsync": Object {
    "hasError": true,
    "isLoading": false,
  },
}
`;

exports[`epics/status should set isLoading property 3`] = `
Object {
  "getVenuesSearchAsync": Object {
    "hasError": false,
    "isLoading": false,
  },
}
`;

exports[`epics/status should set isLoading property 4`] = `
Object {
  "getVenuesSearchAsync": Object {
    "hasError": false,
    "isLoading": false,
  },
}
`;
